graceful/service_signals.go

```diff
@@ -12,40 +11,23 @@ import (
 
 // setupSignals is catching INT/TERM signals to handle a graceful shutdown operation
 // and HUP for a graceful restart. In the case of a restart, the socket is given to the
-// child process to keep receiving new connecions while waiting for the old one to finish
+// child process to keep receiving new connections while waiting for the old one to finish
 // properly.
-func (s *service) setupSignals(ctx context.Context) {
+func (s *Service) setupSignals(ctx context.Context) {
 	log := logger.Get(ctx)
 	ch := make(chan os.Signal, 10)
 	signal.Notify(ch, syscall.SIGINT, syscall.SIGTERM, syscall.SIGHUP)
 	for {
 		sig := <-ch
 		switch sig {
 		case syscall.SIGINT, syscall.SIGTERM:
-			log.Info("request graceful shutdown")
-			signal.Stop(ch)
-			ctx, cancel := context.WithTimeout(ctx, s.waitDuration)
-			defer cancel()
-			err := s.shutdown(ctx)
-			if err != nil {
-				s.stopped <- fmt.Errorf("fail to shutdown service: %v", err)
-			}
-			close(s.stopped)
+			s.graceful.Stop()
 			return
 		case syscall.SIGHUP:
 			log.Info("request graceful restart")
-			_, err := s.graceful.StartProcess()
-			if err != nil {
-				log.WithError(err).Error("fail to start new process")
-			}
-			ctx, cancel := context.WithTimeout(ctx, s.reloadWaitDuration)
-			defer cancel()
-			err = s.shutdown(ctx)
+			err := s.graceful.Upgrade()
 			if err != nil {
-				s.stopped <- fmt.Errorf("fail to shutdown service: %v", err)
-			} else {
-				close(s.stopped)
-				return
+				log.WithError(err).Error("fail to start new service")
```

Comment:
I'm wondering, why did you change the behavior here? (from sending to the `stopped` channel vs logging the error? It's not really clear to me

Reply:
Thank you for pointing this out, I have looked at it again, and will try to explain.

* Previously, there were 2 separate errors; one for `StartProcess` was logged; and one as a result of `shutdown` (of parent process) timing out (sent to the `stopped` channel).
* Now, there is only one error `Upgrade` which is logged. The upgrade does *not* include the termination of the existing process.

Now, the shutdown timeout is handled in file signals.go:142 . If an error occurs or the shutdown times out, the error is returned. This is what the `stopped` channel was used for. This functionality is tested in the test `TestService_Shutdown_WithTimeout`

In fact I should have removed the `stopped` channel, and will push another commit that does so.

This is an important library with regards to the service, so if you think have more comments, please let me know.

Reply:
Your comment suits me, and indeed in this case the channel should be removed